clio-server/src/main/scala/org/broadinstitute/clio/server/service/ServerService.scala

```diff
@@ -55,26 +54,46 @@ class ServerService private (
     */
   private[service] def recoverMetadata(
     implicit index: ElasticsearchIndex[_]
-  ): Future[Int] = {
+  ): Future[Unit] = {
+    val name = index.indexName
+
     searchDAO.getMostRecentDocument.flatMap { mostRecent =>
-      val msg = s"Recovering all upserts from index ${index.indexName}"
+      val msg = s"Recovering all upserts from index $name"
       val latestUpsert = mostRecent.map(getUpsertId)
 
       logger.info(latestUpsert.fold(msg)(id => s"$msg since ${id.id}"))
 
       persistenceDAO
         .getAllSince(latestUpsert)
-        .runWith(Sink.foldAsync(0) { (count, json) =>
-          logger.debug(s"Recovering document with ID ${getUpsertId(json)}")
-          searchDAO.updateMetadata(json).map(_ => count + 1)
-        })
-        .andThen {
-          case Success(count) =>
-            logger.info(s"Recovered $count upserts for index ${index.indexName}")
+        .batch(ServerService.RecoveryMaxBulkSize, json => Map(getEntityId(json) -> json)) {
+          (idMap, json) =>
+            val id = getEntityId(json)
+            val newJson = idMap.get(id) match {
+              case Some(oldJson) => {
+                logger.debug(
+                  s"Merging upserts ${getUpsertId(oldJson)} and ${getUpsertId(json)} for id $id"
+                )
+                oldJson.deepMerge(json)
+              }
+              case None => json
+            }
+            idMap.updated(id, newJson)
         }
+        .runWith(Sink.foldAsync(()) { (_, jsonMap) =>
+          val jsons = jsonMap.valuesIterator.toVector
+          searchDAO.updateMetadata(jsons: _*).map { _ =>
+            logger.debug(s"Sent ${jsons.size} upserts to Elasticsearch for index $name")
```

Comment:
I'll be curious to see how consistent the batch size is when we run in on dev. Can we be sure to output debug logs when we test it?

Reply:
Yup!